### PR TITLE
daemon, Fix unsupported-nics logic in case VF device ID already exists

### PR DESCRIPTION
**Pull request by oshoval**
In case unsupported nic has a VF device id which already exists
in the udev rule, the rule won't be updated, as it depends only on the
VF device ids, and not on the tuple (Vendor, PF_ID, VF_ID).

Therefore:
* Fix logic to update the annotation even if the file contents are the same,
but the annotation is different or doesn't exist at all.
    
Note: since the annotation has the configMap data itself, both on the creation
and validation, it will be robust to configMap updates,
because if the annotation exists but the data in the annotation is different
from the configMap data, it will be rejected, until a new rule will be created,
if needed, and the updated annotation will be written.

Additional problems that were addressed:

* tryCreateUdevRule creates a udev file by using WriteFile
WriteFile will not create the directory hierarchy in case
it doesn't exist [1].
Therefore create the folder in case it doesn't exist.

* In case the daemon updates the unsupported-nic configmap
annotations, before k8s adds `last-applied-configuration` annotation,
a panic will occur at `annotateUnsupportedNicIdConfigMap`:
"Observed a panic: assignment to entry in nil map (assignment to entry in nil map)"
Fix it by allocation the map in case it is still nil.

[1] https://github.com/golang/go/issues/32916

Fixes: https://github.com/k8snetworkplumbingwg/sriov-network-operator/issues/10

Signed-off-by: Or Shoval <oshoval@redhat.com>

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -180,6 +180,15 @@ func (dn *Daemon) getUnsupportedNicIdConfigMap() (*v1.ConfigMap, error) {
 }
 
 func (dn *Daemon) annotateUnsupportedNicIdConfigMap(cm *v1.ConfigMap, nodeName string) (*v1.ConfigMap, error) {
+	annotationKey := "openshift.io/" + nodeName
```
Comment (andreaskaris) [inline pkg/daemon/daemon.go:183]: Can you reuse "annotationKey" in line 200:
    cm.ObjectMeta.Annotations["openshift.io/"+nodeName] = string(jsonData)

Comment (andreaskaris) [reply pkg/daemon/daemon.go:183]: So, change line 200 to:
    cm.ObjectMeta.Annotations["openshift.io/"+nodeName] = string(jsonData)

Comment (oshoval) [reply pkg/daemon/daemon.go]: Done thanks.
Did more little refactors of annotateUnsupportedNicIdConfigMap

**Comment by oshoval** [comment]
```
[Fail] SriovNetwork Controller with SriovNetwork When a derived net-att-def CR is removed [It] should regenerate the net-att-def CR 
/home/runner/work/sriov-network-operator/sriov-network-operator/controllers/sriovnetwork_controller_test.go:204
```
I don't think its due to this PR.
```
2020-12-23T14:15:58.481Z	ERROR	controllers.SriovNetwork	Couldn't update NetworkAttachmentDefinition CR	{"sriovnetwork": "openshift-sriov-network-operator/test-5", "Namespace": "default", "Name": "test-5", "error": "Operation cannot be fulfilled on network-attachment-definitions.k8s.cni.cncf.io \"test-5\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
	/home/runner/work/sriov-network-operator/sriov-network-operator/vendor/github.com/go-logr/zapr/zapr.go:132
github.com/k8snetworkplumbingwg/sriov-network-operator/controllers.(*SriovNetworkReconciler).Reconcile
	/home/runner/work/sriov-network-operator/sriov-network-operator/controllers/sriovnetwork_controller.go:149
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
```
RetryOnConflict would fix it i guess.
Should i retest ?

**Comment by pliurh** [comment]
Retest could work. Additionally, I address this issue by https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/24

**Comment by oshoval** [comment]
Thanks, all tested passed now,
and addressed the comments that were so far.

**Comment by zshi-redhat** [comment]
> In case unsupported nic has a VF device id which already exists
> in the udev rule, the rule won't be updated, as it depends only on the
> VF device ids, and not on the tuple (Vendor, PF_ID, VF_ID).
> 
> Therefore:
> 
>     * Fix logic to update the annotation even if the file contents are the same,
>       but the annotation is different or doesn't exist at all.
> 
> 
> Note: since the annotation has the configMap data itself, both on the creation
> and validation, it will be robust to configMap updates,
> because if the annotation exists but the data in the annotation is different
> from the configMap data, it will be rejected, 

With current PR, if annotation exists but different from the configMap data, it will be updated to configMap data, isn't it?
why it will be rejected?

> until a new rule will be created,
> if needed, and the updated annotation will be written.
> 
> Additional problems that were addressed:
> 
>     * tryCreateUdevRule creates a udev file by using WriteFile
>       WriteFile will not create the directory hierarchy in case
>       it doesn't exist [1].
>       Therefore create the folder in case it doesn't exist.

I guess this issue is not on openshift?

> 
>     * In case the daemon updates the unsupported-nic configmap
>       annotations, before k8s adds `last-applied-configuration` annotation,
>       a panic will occur at `annotateUnsupportedNicIdConfigMap`:
>       "Observed a panic: assignment to entry in nil map (assignment to entry in nil map)"
>       Fix it by allocation the map in case it is still nil.
> 

good catch!

**Comment by oshoval** [comment]
> > Note: since the annotation has the configMap data itself, both on the creation
> > and validation, it will be robust to configMap updates,
> > because if the annotation exists but the data in the annotation is different
> > from the configMap data, it will be rejected,
> 
> With current PR, if annotation exists but different from the configMap data, it will be updated to configMap data, isn't it?

Yes, and until it will be updated it will be rejected, therefore we are safe from races,
because the race can happen if the configMap is updated, and used before we gave the change to update the udev rule.
but since we annotate only after successful udev rule update we are safe.

> why it will be rejected?

Because a preliminary condition in validate.go is that the annotation data equals to the configmap data:
https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/pkg/webhook/validate.go#L318

> 
> > until a new rule will be created,
> > if needed, and the updated annotation will be written.
> > Additional problems that were addressed:
> > ```
> > * tryCreateUdevRule creates a udev file by using WriteFile
> >   WriteFile will not create the directory hierarchy in case
> >   it doesn't exist [1].
> >   Therefore create the folder in case it doesn't exist.
> > ```
> 
> I guess this issue is not on openshift?

I believe you are right, it depends if the folder exists or not there,
we can ask @andreaskaris if he tried that on openshift.
I tried it with nodes that didn't have this folder. 
Anyhow the creation doesn't harm in case the folder exists.
Issue was just that WriteFile is not creating the pull path.

**Comment by zshi-redhat** [comment]
> > > Note: since the annotation has the configMap data itself, both on the creation
> > > and validation, it will be robust to configMap updates,
> > > because if the annotation exists but the data in the annotation is different
> > > from the configMap data, it will be rejected,
> > 
> > 
> > With current PR, if annotation exists but different from the configMap data, it will be updated to configMap data, isn't it?
> 
> Yes, and until it will be updated it will be rejected, therefore we are safe from races,
> because the race can happen if the configMap is updated, and used before we gave the change to update the udev rule.
> but since we annotate only after successful udev rule update we are safe.
> 
> > why it will be rejected?
> 
> Because a preliminary condition in validate.go is that the annotation data equals to the configmap data:
> https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/pkg/webhook/validate.go#L318

Ok. I didn't realize it was about webhook rejection, all good here.
> 
> > > until a new rule will be created,
> > > if needed, and the updated annotation will be written.
> > > Additional problems that were addressed:
> > > ```
> > > * tryCreateUdevRule creates a udev file by using WriteFile
> > >   WriteFile will not create the directory hierarchy in case
> > >   it doesn't exist [1].
> > >   Therefore create the folder in case it doesn't exist.
> > > ```
> > 
> > 
> > I guess this issue is not on openshift?
> 
> I believe you are right, it depends if the folder exists or not on the nodes.
> We can ask @andreaskaris if he tried that on openshift.
> I tried it with nodes that didn't have this folder.
> Anyhow the creation doesn't harm in case the folder exists.
> Issue was just that WriteFile is not creating the pull path.

The change is good, I just to make sure we don't have this issue in openshift.

**Comment by zshi-redhat** [comment]
/lgtm

@pliurh do you want to take a second look?

**Comment by oshoval** [comment]
Thanks @zshi-redhat

**Pull request merged**